src/main/java/org/stratumproject/fabric/tna/behaviour/upf/FabricUpfProgrammable.java

```diff
@@ -645,48 +690,38 @@ public long tableSize(UpfEntityType entityType) throws UpfProgrammableException
+        assertCounterType(type);
         if (!setupBehaviour("readCounter()")) {
             return null;
         }
-        if (cellId >= upfCounterSize || cellId < 0) {
+        if (cellId >= getEntitySize(type) || cellId < 0) {
             throw new UpfProgrammableException("Requested UPF counter cell index is out of bounds.",
                                                UpfProgrammableException.Type.ENTITY_OUT_OF_RANGE);
         }
         UpfCounter.Builder stats = UpfCounter.builder().withCellId(cellId);
 
         // Make list of cell handles we want to read.
-        List<PiCounterCellHandle> counterCellHandles = List.of(
-                PiCounterCellHandle.of(deviceId,
-                                       PiCounterCellId.ofIndirect(FABRIC_INGRESS_UPF_TERMINATIONS_COUNTER, cellId)),
-                PiCounterCellHandle.of(deviceId,
-                                       PiCounterCellId.ofIndirect(FABRIC_EGRESS_UPF_TERMINATIONS_COUNTER, cellId)));
+        List<PiCounterCellHandle> counterCellHandles = Lists.newArrayList();
+        if (isIngressCounter(type) || isBiCounter(type)) {
+            counterCellHandles.add(PiCounterCellHandle.of(
+                    deviceId, PiCounterCellId.ofIndirect(FABRIC_INGRESS_UPF_TERMINATIONS_COUNTER, cellId)
+            ));
+        }
+        if (isEgressCounter(type) || isBiCounter(type)) {
+            counterCellHandles.add(PiCounterCellHandle.of(
+                    deviceId, PiCounterCellId.ofIndirect(FABRIC_EGRESS_UPF_TERMINATIONS_COUNTER, cellId)
+            ));
+        }
 
         // Query the device.
         Collection<PiCounterCell> counterEntryResponse = client.read(
                 DEFAULT_P4_DEVICE_ID, pipeconf)
                 .handles(counterCellHandles).submitSync()
                 .all(PiCounterCell.class);
 
-        // Process response.
-        counterEntryResponse.forEach(counterCell -> {
-            if (counterCell.cellId().counterType() != INDIRECT) {
-                log.warn("Invalid counter data type {}, skipping", counterCell.cellId().counterType());
-                return;
-            }
-            if (cellId != counterCell.cellId().index()) {
-                log.warn("Unrecognized counter index {}, skipping", counterCell);
-                return;
-            }
-            if (counterCell.cellId().counterId().equals(FABRIC_INGRESS_UPF_TERMINATIONS_COUNTER)) {
-                stats.setIngress(counterCell.data().packets(), counterCell.data().bytes());
-            } else if (counterCell.cellId().counterId().equals(FABRIC_EGRESS_UPF_TERMINATIONS_COUNTER)) {
-                stats.setEgress(counterCell.data().packets(), counterCell.data().bytes());
-            } else {
-                log.warn("Unrecognized counter ID {}, skipping", counterCell);
-            }
-        });
-        return stats.build();
+        return processCounterEntryResponse(counterEntryResponse, Lists.newArrayList((long) cellId), type)
```

Comment:
should we handle out-of-bound case ? READ not successful ?

Reply:
you are doing `get(0)` what about if the list is empty. Sorry, I meant this.

Reply:
The list can never be empty, it's initialized at the beginning of the `processCounterEntryResponse` method. Do you suggest to add a check before the return anyway?